src/type.js

```diff
@@ -220,17 +225,28 @@ async function typeImpl(
           newEntry = `-${char}`
         } else if (prevWasPeriod) {
           newEntry = `${prevValue}.${char}`
+        } else if (isValidDateValue(currentElement(), textToBeTyped)) {
+          newEntry = textToBeTyped
         }
 
-        const inputEvent = fireInputEventIfNeeded({
-          ...calculateNewValue(newEntry, currentElement(), currentValue()),
-          eventOverrides: {
-            data: key,
-            inputType: 'insertText',
-            ...eventOverrides,
-          },
-        })
-        prevValue = inputEvent.prevValue
+        if (
+          !(currentElement().type === 'date') ||
+          isValidDateValue(currentElement(), textToBeTyped)
+        ) {
+          const inputEvent = fireInputEventIfNeeded({
+            ...calculateNewValue(newEntry, currentElement(), currentValue()),
+            eventOverrides: {
+              data: key,
+              inputType: 'insertText',
+              ...eventOverrides,
+            },
+          })
+          prevValue = inputEvent.prevValue
+        }
+
+        if (isValidDateValue(currentElement(), textToBeTyped)) {
+          fireEvent.change(currentElement(), {target: {value: textToBeTyped}})
+        }
```

Comment:
[`date` input triggers a `change` event together with `input` event when a date is valid.](https://codesandbox.io/s/user-event-playground-91jlf)